DOC Add secure persistence docs (#145)
skops-dev/skops: docs/persistence.rst

.. warning::

   This feature is very early in development, which means that the API is
   unstable and that it is **not secure** for the moment. Therefore, the same
   caution as for pickle should be applied: Don't load from sources that you
   don't trust. In the future, more security will be added.

Skops offers a way to save and load sklearn models without using :mod:`pickle`.
Pickle is the standard serialization format for sklearn and for Python in
general. It has the big advantage that it can be used for almost all Python code
but this flexibility also means that it's inherently insecure. As the Python
docs say:

    The pickle module is not secure. Only unpickle data you trust. It is
    possible to construct malicious pickle data which will execute arbitrary
    code during unpickling. Never unpickle data that could have come from an
    untrusted source, or that could have been tampered with.

With the :func:`skops.io.save` and :func:`skops.io.load` functions, we aim at
providing a secure method to load and save sklearn code. In contrast to pickle,
these functions cannot be used to save arbitrary Python code, but they bypass
pickle and are thus more secure.

Usage
-----

> Using :func:`skops.io.save` and :func:`skops.io.load` is quite simple. Below is

Reviewer: I'd personally avoid using a "value" word like `simple` or `easy` since it is relative to a user's experience and can make the tone not as friendly for beginners. I think the code snippet below already does a great job of showing how simple it is without having us tell the user it is simple :)

an example:

.. code:: python

    from sklearn.linear_model import LogisticRegression
    from skops.io import load, save

    clf = LogisticRegression(random_state=0, solver="liblinear")
    clf.fit(X_train, y_train)
    save(clf, "my-logistic-regression.skops")
    # ...
    loaded = load("my-logistic-regression.skops")
    loaded.predict(X_test)

At the moment, we support the vast majority of sklearn estimators. This includes
complex use cases such as :class:`sklearn.pipeline.Pipeline`,
:class:`sklearn.model_selection.GridSearchCV`, classes using Cython code, such

> as :class:`sklearn.tree.DecisionTreeClassifier`, etc. If you discover an sklearn

Reviewer: ```suggestion
as :class:`sklearn.tree.DecisionTreeClassifier`, and more. If you discover an sklearn
```

estimator that does not work, please open an issue on the skops `GitHub page
<https://github.com/skops-dev/skops/issues>`_ and let us know.

In contrast to pickle, skops cannot persist arbitrary Python code. In
particular, this means that if you have custom functions (say, a custom function
to be used with :class:`sklearn.preprocessing.FunctionTransformer`), it will not
work. However, most ``numpy`` and ``scipy`` functions should work. Therefore,

> you can actually save ``numpy.sqrt`` etc.

Reviewer 2: ```suggestion
you can actually save built-in functions like``numpy.sqrt``.
```

> In contrast to pickle, skops cannot persist arbitrary Python code. In
> particular, this means that if you have custom functions (say, a custom function
> to be used with :class:`sklearn.preprocessing.FunctionTransformer`), it will not
> work. However, most ``numpy`` and ``scipy`` functions should work. Therefore,
> you can actually save built-in functions like``numpy.sqrt``.

Reviewer: ```suggestion
In contrast to pickle, skops cannot persist arbitrary Python code. This means 
if you have custom functions (say, a custom function to be used 
with :class:`sklearn.preprocessing.FunctionTransformer`), it will not
work. However, most ``numpy`` and ``scipy`` functions should work. Therefore,
you can actually save built-in functions like``numpy.sqrt``.
```

> Goals

Reviewer: Hmm, I'd maybe rename this section to just Roadmap and combine it with the section below.

-----

Currently, it is still possible to run insecure code when using skops

> persistence. E.g., it's possible to load a save file that evaluates arbitrary

Reviewer 2: ```suggestion
persistence, e.g., it's possible to load a save file that evaluates arbitrary
```

> persistence, e.g., it's possible to load a save file that evaluates arbitrary

Reviewer: ```suggestion
persistence. For example, it's possible to load a save file that evaluates arbitrary
```

> code using :func:`eval`. We do, however, have concrete plans on how to mitigate
> this. Please stay updated.

Reviewer: ```suggestion
code using :func:`eval`. However, we have concrete plans on how to mitigate
this, so please stay updated.
```

> On top of trying to support all of sklearn, we plan on making persistence

Reviewer: ```suggestion
On top of trying to support persistence in sklearn, we plan on making persistence
```

.. _persistence:

Secure persistence with skops
=============================

.. warning::

   This feature is very early in development, which means the API is
   unstable and it is **not secure** at the moment. Therefore, use the same
   caution as you would for ``pickle``: Don't load from sources that you

>    don't trust. In the future, more security will be added.

Reviewer 3: we should also tag this as experimental, the same way we do in sklearn, so that it's clear from users' code that this is an experimental feature.

Author: I don't understand how that relates to the documentation. Can we maybe work on this in another PR?

Reviewer 3: Fair enough. Created this: https://github.com/skops-dev/skops/issues/153

Skops offers a way to save and load sklearn models without using :mod:`pickle`.
The ``pickle`` module is not secure, but with skops, you can securely save and
load sklearn models without using ``pickle``.

Pickle is the standard serialization format for sklearn and for Python in

> general. One of its biggest advantages is it can be used for almost all Python

Reviewer 3: The sentence is hard for me to parse, maybe:

```suggestion
general. One of the main advantages of ``pickle`` is that it can be used for almost all Python
```

code but this flexibility also makes it inherently insecure. This is because
loading certain types of objects requires the ability to run arbitrary code,
which can be misused for malicious purposes. For example, an attacker can use it

> to steal secrets from your machine or install a virus. As the Python docs say:

Reviewer 3: link to the python docs?

    The pickle module is not secure. Only unpickle data you trust. It is
    possible to construct malicious pickle data which will execute arbitrary
    code during unpickling. Never unpickle data that could have come from an

>     untrusted source, or that could have been tampered with.

Reviewer 3: maybe put this in a warning tag? (https://sublime-and-sphinx-guide.readthedocs.io/en/latest/notes_warnings.html#warnings)

In contrast to pickle, the :func:`skops.io.save` and :func:`skops.io.load`
functions cannot be used to save arbitrary Python code, but they bypass
pickle and are thus more secure.

Usage
-----

The code snippet below illustrates how to use :func:`skops.io.save` and
:func:`skops.io.load`:

.. code:: python

    from sklearn.linear_model import LogisticRegression
    from skops.io import load, save

    clf = LogisticRegression(random_state=0, solver="liblinear")
    clf.fit(X_train, y_train)
    save(clf, "my-logistic-regression.skops")
    # ...
    loaded = load("my-logistic-regression.skops")
    loaded.predict(X_test)

At the moment, we support the vast majority of sklearn estimators. This includes
complex use cases such as :class:`sklearn.pipeline.Pipeline`,
:class:`sklearn.model_selection.GridSearchCV`, classes using Cython code, such
as :class:`sklearn.tree.DecisionTreeClassifier`, and more. If you discover an sklearn
estimator that does not work, please open an issue on the skops `GitHub page
<https://github.com/skops-dev/skops/issues>`_ and let us know.

> In contrast to pickle, skops cannot persist arbitrary Python code. This means

Reviewer 3: We can probably add support while making it explicit that they are loading arbitrary code.

```suggestion
In contrast to pickle, at the moment skops cannot persist arbitrary Python code. This means
```

Reviewer 3: there might be a way a way to support certain things by using `inspect.getsource`. I'm still not sure what it'd support, but we can have an explicit `arbitrary_code` type which we always warn on.

Author: How about we change the wording once we're there? :)

Reviewer 3: Sure.

.. _persistence:

Secure persistence with skops
=============================

.. warning::

   This feature is very early in development, which means the API is
   unstable and it is **not secure** at the moment. Therefore, use the same
   caution as you would for ``pickle``: Don't load from sources that you
   don't trust. In the future, more security will be added.

Skops offers a way to save and load sklearn models without using :mod:`pickle`.
The ``pickle`` module is not secure, but with skops, you can securely save and
load sklearn models without using ``pickle``.

``Pickle`` is the standard serialization format for sklearn and for Python in
general. One of the main advantages of ``pickle`` is that it can be used for
almost all Python code but this flexibility also makes it inherently insecure.
This is because loading certain types of objects requires the ability to run
arbitrary code, which can be misused for malicious purposes. For example, an

> attacker can use it to steal secrets from your machine or install a virus. As
> the Python docs say:

Reviewer 3: I was thinking something like this would be enough?

```suggestion
attacker can use it to steal secrets from your machine or install a virus. As
the `Python docs <https://docs.python.org/3/library/pickle.html#module-pickle>`__ say:
```

This whole paragraph doesn't have any links to pickle.

Author: This results in the exact same link as ```:mod:`pickle` ``` in the paragraph directly above, but anyway, I added it
